src/client.cpp

```diff
@@ -128,25 +128,69 @@ void Client::initMods()
 
 	// Print mods
 	infostream << "Client Loading mods: ";
-	for (std::vector<ModSpec>::const_iterator i = mods.begin();
-		i != mods.end(); ++i) {
+	for (std::vector<ModSpec>::const_iterator i = m_mods.begin();
+		i != m_mods.end(); ++i) {
 		infostream << (*i).name << " ";
 	}
-
 	infostream << std::endl;
+
 	// Load and run "mod" scripts
-	for (std::vector<ModSpec>::const_iterator it = mods.begin();
-		it != mods.end(); ++it) {
+	for (std::vector<ModSpec>::const_iterator it = m_mods.begin();
+			it != m_mods.end(); ++it) {
 		const ModSpec &mod = *it;
 		if (!string_allowed(mod.name, MODNAME_ALLOWED_CHARS)) {
 			throw ModError("Error loading mod \"" + mod.name +
 				"\": Mod name does not follow naming conventions: "
 					"Only characters [a-z0-9_] are allowed.");
 		}
-		std::string script_path = mod.path + DIR_DELIM + "init.lua";
-		infostream << "  [" << padStringRight(mod.name, 12) << "] [\""
-			<< script_path << "\"]" << std::endl;
-		m_script->loadMod(script_path, mod.name);
+		scanModIntoMemory(mod.name, mod.path);
+	}
+}
+
+void Client::scanModIntoMemory(const std::string &mod_name, const std::string &mod_path)
+{
+	std::vector<fs::DirListNode> mod = fs::GetDirListing(mod_path);
+	for (unsigned int j=0; j < mod.size(); j++){
+		std::string filename = mod[j].name;
+		if (mod[j].dir) {
+			scanModSubfolder(mod_name, mod_path, filename + DIR_DELIM);
+			continue;
+		}
+		m_mod_files[mod_name + ":" + filename] = mod_path + DIR_DELIM + filename;
+	}
+}
+
+void Client::scanModSubfolder(const std::string &mod_name, const std::string &mod_path,
+			std::string mod_subpath)
+{
+	std::string full_path = mod_path + DIR_DELIM + mod_subpath;
+	std::vector<fs::DirListNode> mod = fs::GetDirListing(full_path);
+	for (unsigned int j=0; j < mod.size(); j++){
+		std::string filename = mod[j].name;
+		if (mod[j].dir) {
+			scanModSubfolder(mod_name, mod_path, mod_subpath
+					+ filename + DIR_DELIM);
+			continue;
+		}
+		std::replace( mod_subpath.begin(), mod_subpath.end(), DIR_DELIM_CHAR, '/');
+		m_mod_files[mod_name + ":" + mod_subpath + filename] = full_path + DIR_DELIM + filename;
+	}
+}
```

Comment:
`scanModIntoMemory` and `scanModSubfolder` are very similar and could probably be merged into one function.  `scanModIntoMemory` could just call `scanModSubfolder`, something like `scanModSubfolder(mod_name, mod_path, "")`.